Week 4/Enigma - C Sharp/Enigma/EnigmaMachine.cs

```diff
@@ -139,8 +139,33 @@ public static string CaesarShift(string message, int shift, bool encode)
         /// with the corresponding character in the rotor. </returns>
         private static string ApplyRotor(string message, string rotor)
         {
-            // TO DO - add your implementation
-            throw new NotImplementedException();
+
+            if (rotor.Length != 26)
+                throw new ArgumentException("Rotor must be exactly 26 characters long");
+
+            message = message.ToUpper();
+
+            char[] result = new char[message.Length];
+
+            for (int i = 0; i < message.Length; i++)
+            {
+                char c = message[i];
+
+                // Only substitute letters A-Z
+                if (c >= 'A' && c <= 'Z')
+                {
+                    // Find index in the alphabet (A=0, B=1, ..., Z=25)
+                    int index = c - 'A';
+                    result[i] = rotor[index]; 
+                }
+                else
+                {
+                    // Keep non-alphabetic characters unchanged
+                    result[i] = c; 
+                }
```

Comment:
maybe this could be simplified to:
```suggestion
result[i] = (message[i] - 'A' >= 0 &&  message[i] - 'A' <= rotor.Length) ? rotor[index] : message[i];
```
What do you think?